crates/node/src/types/transaction.rs

```diff
@@ -231,10 +253,42 @@ impl Payload {
 pub enum TransactionError {
     #[error("validation: {0}")]
     Validation(String),
+    #[error("General error: {0}")]
+    General(String),
+}
+
+// Transaction definition.
+// Type state are use to define the different state of a Tx.
+//
+// Tx are defined in 3 domains: Validation, Execution, Storage.
+// Currently the same definition is used but different type should be defined  (TODO).
+// Only the validation type state are defined.
+// TxCreate : identify a Tx that has just been created.
+// TxReceive: Identify a Tx that has been received. Determine the received source.
+// TxValdiated: Identify a Tx that has been validated. Pass all the validation process.
+// The validation suppose the Tx has been propagated. Currently there's no notification during the propagation (TODO).
+
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub struct TxCreate;
+
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub enum TxReceive {
+    P2P,
+    RPC,
+    TXRESULT,
+}
+
+impl TxReceive {
+    fn is_from_tx_exec_result(&self) -> bool {
+        matches!(self, TxReceive::TXRESULT)
+    }
 }
```

Comment:
```suggestion

mod state_domain {
// Transaction definition.
// Type state are use to define the different state of a Tx.
//
// Tx are defined in 3 domains: Validation, Execution, Storage.
// Currently the same definition is used but different type should be defined  (TODO).
// Only the validation type state are defined.
// TxCreate : identify a Tx that has just been created.
// TxReceive: Identify a Tx that has been received. Determine the received source.
// TxValdiated: Identify a Tx that has been validated. Pass all the validation process.
// The validation suppose the Tx has been propagated. Currently there's no notification during the propagation (TODO).

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
pub struct Created;

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
pub enum Received {
    P2P,
    RPC,
    TXRESULT,
}

impl Received {
    fn is_from_tx_exec_result(&self) -> bool {
        matches!(self, TxReceive::TXRESULT)
    }
}
}
```

I wonder if there's any way to implement this same concept using enum? I understand that it doesn't work directly, because Rust doesn't allow generic type definition for enum variant, but it would be a lot more explicit and correct if this was modeled after:
```rust
pub enum TxState {
    Created,
    Received(TxSource),
    Validated,
}

pub enum TxSource {
    P2P,
    RPC,
    TxExecution,
}
```

That way it would be explicit that the transaction lifecycle has certain states and it can be only one of them.

Now the type definition doesn't really communicate this well.

Do you know any pattern or way to implement this without making it look cumbersome (like struct per enum variant, which would be effectively the same as this)?

At the very least, I think putting these types in their own module `state_domain` and using more explicit, but direct naming, would help conveying the idea of this construct to the reader.

Note that this suggestion is not fully function, but just here to express my thinking. :slightly_smiling_face:

Reply:
We can use enum, but the issue is the Rust doesn't currently support [enum variant as type](https://github.com/rust-lang/rfcs/pull/2593) and the RFC has been postponed, so it's not a priority for the Rust community.
You can refer to this [post](https://adjoining-freckle-ff8.notion.site/Type-State-pattern-in-Rust-506aff7f5ce046cd82f16ba6104e09a0) about type state pattern.
The idea to type state is to encode all the state logic in the type, so the compiler can verify we don't do the wrong thing. By using enum you don't encode the logic in the type but in the code, so it's less secure.
There's some drawback in type state so it can't be always use easily and in this case I use enum. Mostly because it's hard to encode in a static way  (the compiler verify) dynamic behaviors.  
For example, for [TxReceive](https://github.com/gevulotnetwork/gevulot/blob/68c879e4df419f5eaebc79170ece38310c34e308/crates/node/src/types/transaction.rs#L275) I use an enum but to be sure there's no bad use I wrap it in a more closed interface using the [TxSender](https://github.com/gevulotnetwork/gevulot/blob/68c879e4df419f5eaebc79170ece38310c34e308/crates/node/src/txvalidation/mod.rs#L58C12-L58C25) type state.
I do that because I didn't want to create a Sink (sender)  for each source (it's can be an option) and as they share the same channel, you can't send different types on one channel. Here, it wasn't too complex to adapt. When you add more complexity, it's better to use enum. Another interesting [article](https://blog.yoshuawuyts.com/state-machines-3/) about that.
One point that is important, I think. Rust is a static language, and it's important to keep as much as possible your code static (everything is validated at compile time). That why I avoid as much as possible to use Dyn trait.

I think we don't have some complex state transition, and we should differentiate them per domain. For example, here I change the Transaction that is share by all domain. Each domain (txvalidation, txexecution, ...)  should have its own transaction type with their own state transition.

Reply:
I change the Tx... naming.